iot-identity/service/service.go

```diff
@@ -87,41 +136,104 @@ func (id IdentityService) EnrollDevice(req *EnrollDeviceRequest) (*domain.Enroll
 		enroll.StoreID = req.Model.Header("store").(string)
 	}
 
-	return id.enroll(&enroll)
+	return id.enroll(&enroll, req.Model, req.Serial)
 }
 
 // Enroll connects an IoT device with the service
-func (id IdentityService) enroll(enroll *datastore.DeviceEnrollRequest) (*domain.Enrollment, error) {
-	// Get the registration for the device
+func (id IdentityService) enroll(enroll *datastore.DeviceEnrollRequest, model asserts.Assertion, serial asserts.Assertion) (*domain.Enrollment, error) {
+	autoRegistrationEnabled := viper.GetBool(keys.AutoRegistrationEnabled)
+
+	log.Infof("Auto-registration is enabled = %t", autoRegistrationEnabled)
+
+	// Get the enrollment for the device, this will exist whether the device itself
+	// has enrolled or not, when a device is registered this is partially created with that info
 	dev, err := id.DB.DeviceGet(enroll.Brand, enroll.Model, enroll.SerialNumber)
-	if err != nil {
-		Logger.Println("Cannot find registration:", err)
-		return nil, err
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		log.Error("Trying to get device: ", err)
+		return nil, fmt.Errorf("getting device: %w", err)
 	}
 
-	// Check that the device is not already enrolled
-	switch dev.Status {
-	case models.StatusWaiting:
-		break
-	case models.StatusEnrolled:
-		return nil, fmt.Errorf("the device `%s/%s/%s` is already enrolled", enroll.Brand, enroll.Model, enroll.SerialNumber)
-	case models.StatusDisabled:
-		return nil, fmt.Errorf("the device registration for `%s/%s/%s` is disabled", enroll.Brand, enroll.Model, enroll.SerialNumber)
-	default:
-		return nil, fmt.Errorf("the device registration for `%s/%s/%s` is invalid", enroll.Brand, enroll.Model, enroll.SerialNumber)
+	if dev != nil && err == nil {
+		log.Tracef("Handling existing device: %s, status is: %d", dev.Device.SerialNumber, dev.Status)
+		switch {
+		case dev.Status == models.StatusWaiting:
+			// this will result in the device being created before function returns
+			break
+		case dev.Status == models.StatusEnrolled:
+			return nil, fmt.Errorf("`%s/%s/%s` is already enrolled", enroll.Brand, enroll.Model, enroll.SerialNumber)
+		case dev.Status == models.StatusDisabled:
+			return nil, fmt.Errorf("`%s/%s/%s` is disabled", enroll.Brand, enroll.Model, enroll.SerialNumber)
+		}
+	} else {
+		log.Tracef("Not existing device, check if auto-registration is enabled and if so, try to register")
+		if errors.Is(err, sql.ErrNoRows) && autoRegistrationEnabled {
```

Comment:
I feel like it's more idiomatic to handle the error conditions first, then we  don't need to guard `dev` against being nil

Reply:
I disagree that it is or is not idiomatic one way or the other.